### DISCUSSION TITLE
Will Struct and Array (like in BigQuery) data types be supported in the future?

### DISCUSSION THREAD

**Discussion by ingted** [NTDLS/Katzebase#66] [Ideas]
As title...

[array_type](https://cloud.google.com/bigquery/docs/reference/standard-sql/data-types#array_type)
[struct_type](https://cloud.google.com/bigquery/docs/reference/standard-sql/data-types#struct_type)

**Comment by ingted** [comment]
Hi @NTDLS,

I am currently doing an experiment to implement array & struct in a cell (field) but when I come to this section, I am confused...

```
public void InsertValue(string fieldNameForException, int ordinal, string? value)
{
    if (Values.Count <= ordinal)
    {
        int difference = (ordinal + 1) - Values.Count;
        if (difference > 0)
        {
            Values.AddRange(new string[difference]);
        }
    }
    if (Values[ordinal] != null)
    {
        throw new KbEngineException($"Ambiguous field [{fieldNameForException}].");
    }

    Values[ordinal] = value;
}
```
Why couldn't it be something like this

```
        public void InsertValue(string fieldNameForException, int ordinal, fstring? value)
        {
            if (Values.Count <= ordinal)
            {
                int difference = (ordinal + 1) - Values.Count;
                if (difference > 0)
                {
                    var r = new string[difference];
                    Values.AddRange(fstring.fromStringArr(r));
                }
            }
            //if (Values[ordinal] != null)
            //{
            //    throw new KbEngineException($"Ambiguous field [{fieldNameForException}].");
            //}

            //Values[ordinal] = value;
        }
```
Just add range when  (difference > 0)?

**Reply by NTDLS** [reply]
There, we are basically building a template row. Depending on the schema that is being joined.


For example, given the query:

`Select b.field1, b.field2, a.field3, a.field4 FROM SchemaA as A INNER JOIN SchemaB.....
`

When we are first scanning for rows in values in SchemaA, we would have values for field3 and field4, but not yet for Field1 and Field2.

The code you were referring to is meant to insert the values for field3 and Field4 and instead placeholders for Field1 and Field2.

**Comment by ingted** [comment]
Keep tracking in [Discussion 71](https://github.com/NTDLS/NTDLS.Katzebase.Server/discussions/71)
